Google.Api.Gax.Grpc/ClientBuilder.cs

```diff
@@ -0,0 +1,205 @@
+        /// invoker is specified.
+        /// </summary>
+        protected virtual async Task<CallInvoker> CreateCallInvokerAsync(CancellationToken cancellationToken)
+        {
+            if (CallInvoker != null)
+            {
+                return CallInvoker;
+            }
+            var endpoint = Endpoint ?? GetDefaultEndpoint();
+            var credentials = await GetChannelCredentialsAsync(cancellationToken).ConfigureAwait(false);
+            Channel channel = new Channel(endpoint.Host, endpoint.Port, credentials);
+            return new DefaultCallInvoker(channel);
+        }
+
+        /// <summary>
+        /// Obtains channel credentials synchronously. Override this method in a concrete builder type if more
+        /// credential mechanisms are supported.
+        /// </summary>
+        protected virtual ChannelCredentials GetChannelCredentials()
+        {
+            if (ChannelCredentials != null)
+            {
+                return ChannelCredentials;
+            }
+            GoogleCredential unscoped =
+                CredentialsPath != null ? GoogleCredential.FromFile(CredentialsPath) :
+                JsonCredentials != null ? GoogleCredential.FromJson(JsonCredentials) :
+                GoogleCredential.GetApplicationDefault();
+            return unscoped.CreateScoped(Scopes ?? GetDefaultScopes()).ToChannelCredentials();
+        }
+
+        /// <summary>
+        /// Obtains channel credentials asynchronously. Override this method in a concrete builder type if more
+        /// credential mechanisms are supported.
+        /// </summary>
+        protected async virtual Task<ChannelCredentials> GetChannelCredentialsAsync(CancellationToken cancellationToken)
+        {
+            if (ChannelCredentials != null)
+            {
+                return ChannelCredentials;
+            }
+            GoogleCredential unscoped =
+                CredentialsPath != null ? GoogleCredential.FromFile(CredentialsPath) : // TODO: Use an async method when one is available
+                JsonCredentials != null ? GoogleCredential.FromJson(JsonCredentials) :
+                await GoogleCredential.GetApplicationDefaultAsync().ConfigureAwait(false);
+            return unscoped.CreateScoped(Scopes ?? GetDefaultScopes()).ToChannelCredentials();
+        }
+
+        /// <summary>
+        /// Returns the default scopes for this builder type, used if no scopes are otherwise specified.
+        /// </summary>
+        protected abstract IReadOnlyList<string> GetDefaultScopes();
+
+        /// <summary>
+        /// Returns the endpoint for this builder type, used if no endpoint is otherwise specified.
+        /// </summary>
+        protected abstract ServiceEndpoint GetDefaultEndpoint();
+
+        /// <summary>
+        /// Builds the resulting client.
+        /// </summary>
+        /// <remarks>The implementation is responsible for performing validation before constructing the client.</remarks>
```

Comment:
nit: consider adding "by calling the Validate method" or similar.

Reply:
I wonder whether I should remove the remarks entirely, in fact. It's only relevant to implementers, which we expect to be just us. Users shouldn't care.

Reply:
Well, a reminder for us is not a waste I would say. The other implementation with Build and BuildImpl was harder to mess with by forgetting to call Validate, but here is easier. I like this better because is cleaner, but I think the comment has its value.

Reply:
It definitely has value, but probably not in terms of generated documentation. Just a `//` comment instead?

Reply:
Ah yes, that's entirely fine.